Adding in handbook the production readiness review (#1219)
mattermost/mattermost-handbook: operations/research-and-development/engineering/infrastructure-engineering/reliability-manifesto.md

> # Reliability Manifesto

Reviewer: While perhaps not the usual category in a doc like this, I'd like to have a "transparency" manifesto with the goal of increasing trust and observability, e.g.:
* Repositories should be default open, including to the wider public, and hosted on GitHub.
* When the repository contains details of our internal systems (e.g. infrastructure not meant to be shared publicly), repositories should be accessible to all staff with VPN access, and hosted on the internal GitLab.
* In rare cases, some repositories may need to be further constrained (e.g. security). Seek an exception to create such a private repository.

Author: +1 though I don't believe is part of this doc. right?

**F-1** We do not silently bypass the rules in this document, but instead, start a discussion to change the rules when we think something does not fit our situation.

**F-2** This set of rules should reflect measurable targets in the quarter OKRs of the teams that utilize the Infrastructure platform. 

**F-3** We assess the validity of the rules in this document twice a year and we adapt accordingly.

## Continuous Improvement

There is always room for improvement in our knowledge and our systems. The continuous improvement manifesto aims to make us more efficient, resilient and better engineers.

**C-1** We should track all our production incidents. Major and critical incidents are reported and updated using our statuspage. All incidents are tracked via a Infrastructure Incident Response v2.0 playbook run. More details here.

**C-2** Post mortem meetings and documents should follow every major and critical production incident. Runbooks if any should be added in the internal documentation for future reference.

**C-3** No single engineer should fully deploy a new service across all environments. We share knowledge and reduce the team bus factor. 

> **C-4** Blameless culture is a key. People make mistakes and systems (do) fail. We need to learn from them and help each other. 

Reviewer: ```suggestion
**C-4** Blameless culture is a key. People make mistakes and systems do fail. We need to learn from them and help each other.
```

## Design & Architecture

Better design and architectures create more reliable systems and reduce technical debt. This is what the Design & Architecture manifesto is all about. 

**D-1** Infrastructure team should be involved early in the design process for applications that will run in the infrastructure systems. Read about the [Production Readiness Review](/operations/research-and-development/engineering/infrastructure-engineering/production-readiness-review.md)

> **D-2** All new services should go through the POC phase and findings should be presented in a team meeting, together with proposal, design architecture and implementation steps. Example reference [here](https://docs.google.com/document/d/1KZlemVB_75Xv4CDao5bcWpKml3fHbpcoAk_gWl8nFSY/edit).

Reviewer 2: Thanks! Docs all good to me but left one suggestion to expand the "Testing" section into "Testing and Qualification".

Author: I will update accordingly Saturn, thanks

> **D-2** All new services should go through the POC phase and findings should be presented in a team meeting, together with proposal, design architecture and implementation steps. Example reference [here](https://docs.google.com/document/d/1KZlemVB_75Xv4CDao5bcWpKml3fHbpcoAk_gWl8nFSY/edit).

Reviewer 3: I added a comment to the design doc template after discussing with Saturn.

> **D-3** We should avoid monolithic designs for custom services when possible. Microservice design should be the aim and services should be open sourced for community usage when they don’t expose private company information.

Reviewer: The microservice / monolith boundary can be blurry, and I'm partial to the "microlith" approach vs. what the industry previously blessed as "pure microservice". Not sure we need to overthink the text here, but just thinking out loud.

Author: Agree i think we can revisit this. @stylianosrigas

> **D-4** All services should be followed by architecture diagrams (we use Lucidchart) when possible and documentation. 

Reviewer: Can we say, "e.g. Lucidchart" vs what sounds like a prescription to use Lucidchart? I use Monodraw ;P
```suggestion
**D-4** All services should be followed by architecture diagrams (e.g. Lucidchart) when possible and documentation.
```

**D-5** Always be mindful that services work in a different way at scale. Configurations that might work in local environments could have a major impact in production systems. For example, a possible abuse of the amount of database connections used. 

**D-6** All services should support the ability to turn off/on on demand without affecting the rest of the infrastructure services.

## Observability

The observability manifesto should cover three main pillars:
- Logs
- Metrics
- Tracing

> ### Logs

Reviewer: Something I don't see here is an opinion on levels. I'm of the mind that ERROR logs should be human actionable, otherwise they are at best warnings. Also, our INFO level is surprisingly sparse: all the good stuff is at DEBUG.

Reviewer 4: Do we have well defined definitions for what an "INFO" event is and when it should be placed there?  I guess the other question would be: Is there value in separating these levels out and just having a single application log firehose and an audit log one?

Author: we had in the past a section for it and I am ok if we want to add it. could you propose a change @lieut-data

Reviewer: Sounds good -- proposed something below, and looped in @agnivade for opinions on that.

Reviewer 5: This is our guidance on log levels: https://developers.mattermost.com/contribute/more-info/server/style-guide/#log-levels.

Reviewer 4: Linking to the developer docs gives us the single source of truth and we don't need to modify in multiple places if it ever changes. They look good to me!

**O-1** Logs should include meaningful messages and stack trace should not be a requirement to understand the log message.

> **O-2** Service logs are enriched with metadata & attributes that differentiate the service (e.g service: boards).

Reviewer: Propose a different sample service:
```
**O-2** Service logs are enriched with metadata & attributes that differentiate the service (e.g service: ldap).
```

> **O-3** Logging levels should be consistent across services and false errors should be avoided to enable generation of metrics, patterns and dashboards.

Reviewer: ```suggestion
**O-3** Logging levels should be consistent across services and false errors should be avoided to enable generation of metrics, patterns and dashboards. By default, applications will log at `INFO` and above and produce a relatively low volume of logs.
- Use `ERROR` when a human should take action to remediate a failure such as database connection issues, incorrect cluster configuration, inability to reach a first-party host (e.g. S3, rtcd) or an application segfault.
- Use `WARN` for something unexpected, but typically recoverable without human intervention, such as the inability to reach a third-party host (e.g. GitHub, Giphy), an unexpected payload, or a failed client API call. Assume that infrastructure tooling will monitor for a spike in warning logs to trigger human action as needed.
- Use `INFO` to capture context around non-user initiated events such as startup information, plugin state changes, schema migrations, background jobs starting and finishing, or configuration changes. 
- Use `DEBUG` to trace user-initiated events or add context to other systems such as client API requests, the details of the LDAP Sync job or cluster polling.
```

cc @agnivade for opinions on all this

Reviewer 5: We should just point to: https://developers.mattermost.com/contribute/more-info/server/style-guide/#log-levels.

Author: @lieut-data let's add the link as proposal maybe then?

Reviewer: Oh, goodness, I forgot about that altogether! Multiple sources of truth :/ -- I'm fine with doing nothing here.

Author: I would say let's just refer the link.

Reviewer 4: Agreed, developer docs is good for me!

**O-4** Logging format structure should be consistent across services and should be agreed across all teams. String and JSON formats should be used for all logs. 

**O-5** Retention of 30 days for all critical and 7 days of all non-critical production system logs should be met. With exceptions, logs can be kept up to 1 year in S3 utilizing Glacier options.

### Metrics
**O-6** All services should export key metrics under /metrics path and should be enriched with metadata & attributes that differentiate the service (e.g service: playbooks).

**O-7** All the application metrics should be exposed via 8067 port.

**O-8** Application and service software version metrics of all applications should be exposed.

**O-** The application and services should expose availability and latency metrics.

**O-10** Raw metrics should be kept for 7 days and 5m/1h resolution metrics should be kept for 365 days for all production systems. 

### Tracing (Placeholder)

> O-11 As we are moving more into microservices, request tracing should be implemented for each microservice to enable distributed request tracing analysis.

Reviewer 2: Could you share some context or reference about moving to microservices? I'm specifically interested to know the scope/breakdown of app/services to be moved.

Author: @saturninoabril  This is mostly applicable for the Mattermost Cloud, I would say which we need to revisit it a bit. For example:
- CWS = microservice
- Provisioner = microservice
- Elrond is a microservice
- Database factory is a microservice

etc. etc.

Reviewer 2: Got it, thanks!

> **O-11** As we are moving more into microservices, request tracing should be implemented for each microservice to enable distributed request tracing analysis.

Reviewer: This seems highly aspirational? I'm ok with that, but if some of this is "future thinking" but not realistic now, I worry it's easy to ignore other parts of the doc.

Author: we can remove this, I agree.
